File: GUI Objects/FoodTableDragSupport.m
#import "FoodTableDragSupport.h"
#import "FoodDragSupportBase.h"
#import "BeFit-Swift.h"

@implementation FoodTableDragSupport

//Need to implement data source routines so we include them
- (int)numberOfRowsInTableView:(NSTableView *)aTableView 
{
	mDisplayedTableView = aTableView;
	return 0; // return 0 so the table view will fall back to getting data from its binding
}

- (id)tableView:(NSTableView *)aTableView objectValueForTableColumn:(NSTableColumn *)aTableColumn row:(int)rowIndex 
{
	return nil;  // return nil so the table view will fall back to getting data from its binding 
}

- (NSTableView *)currentlyDisplayedTableView
{
	return mDisplayedTableView;
}

- (BOOL)tableView:(NSTableView*)tableView 
        writeRows:(NSArray*)rows 
        toPasteboard:(NSPasteboard*)pboard
{
    // Get array controller
    NSDictionary*       bindingInfo;
    NSArrayController*  arrayController;
    bindingInfo = [tableView infoForBinding:NSContentBinding];
    arrayController = [bindingInfo valueForKey:NSObservedObjectKey];
    
    // Get arranged objects, they are managed object
    NSArray*    arrangedObjects;
    arrangedObjects = [arrayController arrangedObjects];
    
    // Collect URI representation of managed objects
    NSMutableArray* objectURIs;
    NSEnumerator*   enumerator;
    NSNumber*       rowNumber;
    objectURIs = [NSMutableArray array];
    enumerator = [rows objectEnumerator];
    while (rowNumber = [enumerator nextObject]) {
        int row;
        row = [rowNumber intValue];
        if (row >= [arrangedObjects count]) {
            continue;
        }
        
        // Get URI representation of managed object
        NSManagedObject*    object;
        NSManagedObjectID*  objectID;
        NSURL*              objectURI;
        object = [arrangedObjects objectAtIndex:row];
        objectID = [object objectID];
        objectURI = [objectID URIRepresentation];
        
        [objectURIs addObject:objectURI];
    }
    
    // Set them to paste board
    [pboard declareTypes:[NSArray arrayWithObject:FoodDataDragType] owner:nil];
    [pboard setData:[NSArchiver archivedDataWithRootObject:objectURIs] forType:FoodDataDragType];
    
    return YES;
}

@end

File: GUI Objects/FoodTableDragSupport.swift
//
//  FoodTableDragSupport.swift
//  BeFit
//
//  Created by Andrej Slegl on 11/5/19.
//

import Foundation
import Cocoa

class FoodTableDragSupport: NSObject, NSTableViewDataSource, NSTableViewDelegate {
    private var mDisplayedTableView: NSTableView?
    
    var currentlyDisplayedTableView: NSTableView? {
        return mDisplayedTableView
    }
    
    func numberOfRows(in tableView: NSTableView) -> Int {
        mDisplayedTableView = tableView
        return 0 // return 0 so the table view will fall back to getting data from its binding
    }
    
    func tableView(_ tableView: NSTableView, objectValueFor tableColumn: NSTableColumn?, row: Int) -> Any? {
        return nil // return nil so the table view will fall back to getting data from its binding
    }
    
    func tableView(_ tableView: NSTableView, writeRowsWith rowIndexes: IndexSet, to pboard: NSPasteboard) -> Bool {
        guard let bindingInfo = tableView.infoForBinding(.content) else {
            return false
        }
        
        guard let arrayController = bindingInfo[.observedObject] as? NSArrayController else {
            return false
        }
        
        let arrangedObjects = arrayController.arrangedObjects as! [NSManagedObject]
        
        // Collect URI representation of managed objects
        let objectURIs = rowIndexes.compactMap { row -> URL? in
            guard row < arrangedObjects.count else {
                return nil
            }
            
            let object = arrangedObjects[row]
            return object.objectID.uriRepresentation()
        }
        
        pboard.declareTypes([.foodDataDragType], owner: nil)
        pboard.setData(NSArchiver.archivedData(withRootObject: objectURIs), forType: .foodDataDragType)
        
        return true
    }
}
